src/community.c

```diff
@@ -3567,3 +3567,1094 @@ int igraph_i_compare_communities_rand(const igraph_vector_t *v1,
+ * the cluster of node i and d(x, y) = 1 if and only if x = y and 0 otherwise.
+ * By setting n_i = k_i, the degree of node i, and dividing gamma by 2m, you
+ * effectively obtain an expression for modularity. Hence, the standard
+ * modularity will be optimized when you supply the degrees as \c node_weights
+ * and by supplying as a resolution parameter 1.0/(2*m), with m the number of
+ * edges.
+ *
+ * \param graph The input graph. It must be an undirected graph.
+ * \param edge_weights Numeric vector containing edge weights. If \c NULL, every edge
+ *    has equal weight of 1. The weights need not be non-negative.
+ * \param node_weights Numeric vector containing node weights.
+ * \param resolution_parameter The resolution parameter used, which is
+ *    represented by gamma in the objective function mentioned in the
+ *    documentation.
+ * \param beta The randomness used in the refinement step when merging. A small
+ *    amount of randomness (beta = 0.01) typically works well.
+ * \param membership The membership vector. This is both used as the initial
+ *    membership from which optimisation starts and is updated in place. It
+ *    must hence be properly initialized. When finding clusters from scratch it
+ *    is typically started using a singleton clustering. This can be achieved
+ *    using \c igraph_vector_init_seq.
+ * \param nb_clusters The number of clusters contained in \c membership. Must
+ *    not be a \c NULL pointer.
+ * \param quality The quality of the partition, in terms of the objective
+ *    function as included in the documentation. If \c NULL the quality will
+ *    not be calculated.
+ * \return Error code.
+ *
+ * Time complexity: near linear on sparse graphs.
+ *
+ * \example examples/simple/igraph_community_leiden.c
+ */
+int igraph_community_leiden(const igraph_t *graph,
+  const igraph_vector_t *edge_weights, const igraph_vector_t *node_weights,
+  const igraph_real_t resolution_parameter, const igraph_real_t beta,
+  igraph_vector_t *membership, igraph_integer_t *nb_clusters, igraph_real_t *quality) 
+{
+  igraph_vector_t *i_edge_weights, *i_node_weights;
+  int ret;
+
+  if (igraph_vector_size(membership) != igraph_vcount(graph)) 
+    IGRAPH_ERROR("Membership length does not equal the number of vertices", IGRAPH_EINVAL);
+
+  if (igraph_is_directed(graph))
+    IGRAPH_ERROR("Leiden algorithm is only implemented for undirected graphs", IGRAPH_EINVAL);
+    
+  /* Check edge weights to possibly use default */
+  if (!edge_weights)
+  {
+      i_edge_weights = igraph_Calloc(1, igraph_vector_t);
+      if (i_edge_weights == 0)
+        IGRAPH_ERROR("Leiden algorithm failed, could not allocate memory for edge weights", IGRAPH_ENOMEM);    
+      IGRAPH_CHECK(igraph_vector_init(i_edge_weights, igraph_ecount(graph)));
+      IGRAPH_FINALLY(free, i_edge_weights);
+      IGRAPH_FINALLY(igraph_vector_destroy, i_edge_weights);
+      igraph_vector_fill(i_edge_weights, 1);
```

Comment:
Minor comment: The `igraph_vector_t` `struct` could have been allocated on the stack for simplicity, saving some `free`/`calloc`. (Of course, the actual vector storage still needs to be allocated/freed conditionally on whether the user supplied any weights).

Reply:
That might be possible. However, I am not sure how I then do
```
i_edge_weights = edge_weights;
```
in case that `i_edge_weights` is allocated on the stack?

Reply:
Something going wrong with GitHub. My comment triggered a new review by me, which I then deleted, but it also deleted the original comment.

It might be possible to use the stack instead of the heap, but I am not sure how to do
```
i_edge_weights = edge_weights;
```
in that case.

Reply:
We could do

```
igraph_vector_t i_edge_weights_vec;
igraph_vector_t *i_edge_weights;
```

Then set `i_edge_weights = &i_edge_weights_vec;` or `i_edge_weights = edge_weights;` as needed.

But again, it's not a big deal.

Reply:
Thanks, that would be a possibility indeed. But I think I then prefer the current solution.